resilience4j-rxjava2/src/main/java/io/github/resilience4j/retry/transformer/RetryTransformer.java

```diff
@@ -58,118 +41,78 @@ private RetryTransformer(Retry retry) {
-        });
+        //noinspection unchecked
+        Context<T> context = new Context<T>(retry.context());
+        return upstream.doOnNext(context::throwExceptionToForceRetryOnResult)
+                .retryWhen(errors -> errors.doOnNext(context::onError))
+                .doOnComplete(context::onComplete);
     }
 
     @Override
     public SingleSource<T> apply(Single<T> upstream) {
-        return Single.fromPublisher(downstream -> {
-            Flowable<T> flowable = upstream.toFlowable();
-            SubscriptionArbiter sa = new SubscriptionArbiter();
-            downstream.onSubscribe(sa);
-            RetrySubscriber<T> retrySubscriber = new RetrySubscriber<>(downstream, retry.getRetryConfig().getMaxAttempts(), sa, flowable, retry);
-            flowable.subscribe(retrySubscriber);
-        });
+        //noinspection unchecked
+        Context<T> context = new Context<T>(retry.context());
+        return upstream.doOnSuccess(context::throwExceptionToForceRetryOnResult)
+                .retryWhen(errors -> errors.doOnNext(context::onError))
+                .doOnSuccess(t -> context.onComplete());
     }
 
-    static final class RetrySubscriber<T> extends AtomicInteger implements Subscriber<T> {
-
-        private final Subscriber<? super T> actual;
-        private final SubscriptionArbiter sa;
-        private final Publisher<? extends T> source;
-        private final Retry.Context context;
-        private long remaining;
-        RetrySubscriber(Subscriber<? super T> actual, long count,
-                         SubscriptionArbiter sa, Publisher<? extends T> source,
-                         Retry retry) {
-            this.actual = actual;
-            this.sa = sa;
-            this.source = source;
-            this.context = retry.context();
-            this.remaining = count;
-        }
+    @Override
+    public CompletableSource apply(Completable upstream) {
+        //noinspection unchecked
+        Context<T> context = new Context<T>(retry.context());
+        return upstream.retryWhen(errors -> errors.doOnNext(context::onError))
+                .doOnComplete(context::onComplete);
+    }
 
-        @Override
-        public void onSubscribe(Subscription s) {
-            if (LOG.isDebugEnabled()) {
-                LOG.info("onSubscribe");
-            }
-            sa.setSubscription(s);
+    @Override
+    public MaybeSource<T> apply(Maybe<T> upstream) {
+        //noinspection unchecked
+        Context<T> context = new Context<T>(retry.context());
+        return upstream.doOnSuccess(context::throwExceptionToForceRetryOnResult)
+                .retryWhen(errors -> errors.doOnNext(context::onError))
+                .doOnSuccess(t -> context.onComplete())
+                .doOnComplete(context::onComplete);
+    }
+
+    private static class Context<T> {
+        private static final RuntimeException RESULT_EXCEPTION = new RuntimeException("retry due to retryOnResult predicate");
```

Comment:
If we have to use exceptions for flow control, we should introduce a custom runtime exception like
`RetryDueToResultException`. What do you think?

Reply:
Sounds good will do that